#167165061 #167165069 CRUD articles and create tags for articles

tolulope-od opened the pull request:
### What does this PR do?

Adds functionality to create articles and add tags to an article
Modifies database model for articles to add new data types
Modifies server response method
Adds validation when creating an article
Creates a route to get articles by tag
Creates a route to get an article by author and slug
Creates a route for an author to get their own articles
Adds more tags support up to 10
Change server success response format for profiles

#### Description of Task to be completed?
Users should be able to Create, Read, Update, Delete articles (WIP)
Create articles - `POST /api/v1/articles`
Get all published articles - `GET -> /api/v1/articles`
Get articles by a tag - `GET /api/v1/articles?tag=`
Get articles by slug - `GET /api/v1/:email/:slug`
Get all published articles by an author - `GET /api/v1/:email`
Get all articles by an author including drafts (for author) - `GET /api/v1/users/articles`

#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin  ft-user-create-article-and-tags-167165061-167165069

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```
Using postman, log in or register an account and hit the post route to create an article (user must be verified via the verification route in #28 

add the following format in the request body:

```JSON
{
 "title": "How to live a fulfilling life",
  "description": "Do what you love",
  "body": "Lorem Ipsum",
  "tags": ["life", "advice", "career", "self-care"],
  "status": "published"
}
```

#### Any background context you want to provide?
N/a

#### What are the relevant pivotal tracker stories?
[#167165061](https://www.pivotaltracker.com/story/show/167165061) 
[#167165069](https://www.pivotaltracker.com/story/show/167165069)

#### Screenshots (if appropriate)

*Created article successfully*
<img width="1680" alt="Screenshot 2019-08-01 at 08 16 46" src="https://user-images.githubusercontent.com/42325628/62273539-b08fef00-b435-11e9-87ec-fc4874c4b5c1.png">

*Get all published articles by an author*
<img width="1680" alt="Screenshot 2019-08-01 at 08 16 46" src="https://user-images.githubusercontent.com/42325628/62273765-3a3fbc80-b436-11e9-80b2-824fa887e6be.png">

*Get articles by a tag*

<img width="1680" alt="Screenshot 2019-08-01 at 08 19 06" src="https://user-images.githubusercontent.com/42325628/62274024-eed9de00-b436-11e9-8916-ed2e985cdbc9.png">

tolulope-od committed:
feature(articles): create article and add tags

Add functionality to create articles and add tags to an article
Modify database model for articles to add new data types
Modify server response method
Add validation when creating an article

[Starts #167165061, #167165069]

tolulope-od committed:
feature(articles): get articles

Create get articles by tag
Create get article by author and slug
Create author get own articles
Add more tags support up to 10
Change server success response format for profiles

[BuildsOn #167165061]

VictorMaria commented:
Well done Tolu, your tests are comprehensive. You may try rebasing you branch due to the merge conflicts. Good job.

tolulope-od committed:
feature(articles):edit and delete article

Create route to edit an article
Create route to delete an article
Change getting article by author to use ID instead of emails
Document all features with Swagger
Add UUID format validation when fetching any resource requiring UUID

[BuildsOn #167165061]

tolulope-od commented:
> Well done Tolu, your tests are comprehensive. You may try rebasing you branch due to the merge conflicts. Good job.

Thanks Victor, I've fixed rebased the branch

tolulope-od commented:
I want to apologize for having so many diffs on one branch, I tried to break things down into commits to reduce the number of diffs, but the task was quite huge and required a lot of changes and additions.

Tunji17 reviewed (approved):
Great job @tolulope-od , LGTM

mazma1 commented on __tests__/articleRoutes.test.js:
```diff
@@ -0,0 +1,103 @@
+        expect(res.status).to.be.eql(400);
+        expect(res.body).to.have.property('errors');
+        expect(res.body.errors).to.be.to.a('object');
+        expect(res.body.errors).to.have.property('tags').eql('Only a maximum of 3 tags are allowed');
```
Only a maximum of 3 tags are allowed?

tolulope-od replied:
No, this is an outdated commit. It has been increased to 10 now

mazma1 commented on database/models/article.js:
```diff
@@ -27,6 +27,7 @@ module.exports = (sequelize, DataTypes) => {
       },
       ratings: {
         type: DataTypes.INTEGER,
+        defaultValue: 0
```
I don't know how you all manage to get away with no trailing commas shaa

tolulope-od replied:
I try to add it as much as I can, but ESLint flags where it's required, I don't get why it's selective about the flagging sometimes

mazma1 commented:
You did a great work on this @tolulope-od, well done!  🎉 🎉
